clientv3/watch.go

```diff
@@ -456,11 +463,23 @@ func (w *watchGrpcStream) run() {
 					wc.Send(ws.initReq.toPB())
 				}
 			case pbresp.Canceled:
-				delete(cancelSet, pbresp.WatchId)
-				if ws, ok := w.substreams[pbresp.WatchId]; ok {
-					// signal to stream goroutine to update closingc
-					close(ws.recvc)
-					closing[ws] = struct{}{}
+				if len(pbresp.CancelReason) != 0 {
```

Comment:
I tried the idea of setting `Created` as `true`, but from the cancel reason couldn't propagated to a client. This is because `pbresp.WatchId` is -1 so the existing path of `pbresp.Created` cannot reply the watch response to the client. It is because handling the canceled response with the path of created will close `watchGrpcStream.revc` in `watchGrpcStream.addSubstream()`.
 This is why I added the new path and new function `dispatchCanceledEvent()`: https://github.com/coreos/etcd/pull/7809/files#diff-bdef56e462f289cae8e13ebbe16ba455R571

Of course the watching can be canceled but the client cannot know the reason. How do you think?